src/libos/src/events/notifier.rs

```diff
@@ -0,0 +1,62 @@
+use std::any::Any;
+use std::sync::Weak;
+
+use super::{Event, EventFilter, Observer};
+use crate::prelude::*;
+
+// TODO: how to prevent deadlock or infinite loop during event propagation?
+
+/// Event notifier.
+pub struct Notifier<E: Event, F: EventFilter<E>> {
+    subscribers: SgxMutex<VecDeque<Subscriber<E, F>>>,
+}
+
+struct Subscriber<E: Event, F: EventFilter<E>> {
```

Comment:
If no filter is needed, should an `EventFilter` be provided?

Reply:
Theoretically speaking, there could be scenarios where event filters are not needed. But in reality, this is not likely to happen. Even if it happens, a type that implements `Event` usually have a reasonable type that can implement `EventFilter`. For example, `SigSet` is a proper event filter for `SigNum` as an event. I cannot think of an event type that cannot provide a reasonable, companion filter type.

With that said, I added a `DummyEventFilter` type that can serve as a default type for the parameter type `F` of `Notifier<E: Event, F: EventFilter<E> = DummyEventFilter<E>>`. In this way, if an event filter is not needed, then the user can just use `Notifier<E>` without explicitly giving the event filter type.